yapapi/contrib/service/http_proxy.py

```diff
@@ -0,0 +1,173 @@
+        """
+        assert self.network_node, "Ensure that the service is started within a network."
+        assert self.cluster
+
+        instance_ws = self.network_node.get_websocket_uri(self._remote_port)
+        app_key = self.cluster.service_runner._job.engine._api_config.app_key  # noqa
+
+        remote_headers = "\r\n".join(
+            [
+                f"{k}: {v if k != 'Host' else self._remote_host or v}"
+                for k, v in request.headers.items()
+            ]
+        )
+
+        remote_request: bytes = (
+            f"{request.method} {request.path_qs} "
+            f"HTTP/{request.version.major}.{request.version.minor}\r\n"
+            f"{remote_headers}\r\n\r\n"
+        ).encode("ascii") + (await request.read() if request.can_read_body else b"")
+
+        logger.info("Sending request: `%s %s` to %s", request.method, request.path_qs, self)
+        logger.debug("remote_request: %s", remote_request)
+
+        ws_session = aiohttp.ClientSession()
+        async with ws_session.ws_connect(
+            instance_ws,
+            headers={"Authorization": f"Bearer {app_key}"},
+        ) as ws:
+            await ws.send_bytes(remote_request)
+
+            response_parser = _ResponseParser(ws, self._remote_response_timeout)
+            try:
+                response = await response_parser.get_response()
+                logger.info("Remote response received. status=%s", response.status)
+                logger.debug(
+                    "Remote response: status=%s, headers=%s, body=%s",
+                    response.status,
+                    response.headers,
+                    response.body,
+                )
+            except Exception as e:
+                logger.error(
+                    "Error receiving remote response. url=%s, service=%s, exception=%s(%s) [%s])",
+                    request.path_qs,
+                    self,
+                    type(e),
+                    str(e),
+                    traceback.format_exc(),
+                )
+                response = web.Response(status=500, text="Error retrieving the remote response.")
+
+        await ws_session.close()
+
+        return response
+
+
+class LocalHttpProxy:
+    """runs a local aiohttp server and processes requests through instances of HttpProxyService."""
+
+    def __init__(self, cluster: Cluster[HttpProxyService], port: int):
+        self._request_count = 0
+        self._request_lock = asyncio.Lock()
+        self._cluster = cluster
+        self._port = port
+        self._site = None
+
+    async def request_handler(self, request: web.Request):
+        logger.info("Received a local HTTP request: %s %s", request.method, request.path_qs)
+
+        instances = [i for i in self._cluster.instances if i.state == ServiceState.running]
+        async with self._request_lock:
+            instance: HttpProxyService = instances[self._request_count % len(instances)]
```

Comment:
I think I've written this somewhere, but can't find the comment now.

So, again: this **assumes there is any instance running**, but this assumption is not checked and is not really obvious.
We could have here something like:
```
instances = []
while not instances:
    instances = [i for i in self._cluster.instances if i.state == ServiceState.running]
    if not instances:
        logger.debug(...)
```    
or something similar.

IOW, if this is in `contrib`, it **should** be easy to use and as useful as possible.
Additional assumption "use this only with a cluster with at least one running instance" is not obvious, but also not necessary.

Reply:
Found! https://github.com/golemfactory/yapapi/pull/860#discussion_r811398535

Reply:
I have no idea why this comment eluded me ...

Reply:
anyway, fixed, thank you @johny-b